File: frontend/src/features/blog/http/api/UpdateBlog.ts
import type { BlogDetail } from '@/features/blog/model/response/BlogDetail'
import type { UpdateBlogRequest } from '@/features/blog/model/request/UpdateBlogRequest'
import { blogIdValue } from '@/features/blog/lib/blog-parsers'
import {
  fromBlogDetailContract,
  toUpdateBlogRequestContract,
} from '@/features/blog/http/codec/BlogHttpCodecs'
import type { BlogId } from '@/features/blog/model/BlogId'
import { postJson } from '@/shared/api/http-client'

export async function updateBlog(blogId: BlogId, request: UpdateBlogRequest): Promise<BlogDetail> {
  return postJson(`/api/blogs/${blogIdValue(blogId)}/update`, fromBlogDetailContract, toUpdateBlogRequestContract(request))
}

File: backend/src/main/scala/domains/blog/http/api/UpdateBlog.scala
package domains.blog.http.api



import domains.blog.http.*
import domains.blog.http.codec.BlogHttpCodecs.given
import domains.blog.http.mapper.BlogHttpRequestMappers
import cats.effect.IO
import domains.blog.application.BlogCommands
import domains.blog.model.request.UpdateBlogRequest
import domains.blog.model.BlogId
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.Http4sDsl
import org.http4s.dsl.io.*

object UpdateBlog:

  def routes(context: BlogHttpRouteContext)(using Http4sDsl[IO]): HttpRoutes[IO] =
    HttpRoutes.of[IO] {
      case request @ POST -> Root / "api" / "blogs" / rawBlogId / "update" =>
        BlogHttpRequestMappers.blogId(rawBlogId) match
          case Left(message) => shared.http.utils.HttpResponseSupport.validationErrorResponse(message)
          case Right(blogId) =>
            context.handlers.executeDecoded[UpdateBlogRequest, BlogHttpPlans.UpdateBlogInput, BlogCommands.UpdateBlogResult](
              request,
              context.plans.updateBlog
            )(updateRequest => BlogHttpPlans.UpdateBlogInput(blogId, updateRequest))
    }
